src/util.ts

```diff
@@ -0,0 +1,217 @@
+  };
+}
+
+export function urlGenerate(aParsedUrl: ParsedURL) {
+  let url = '';
+  if (aParsedUrl.scheme) {
+    url += aParsedUrl.scheme + ':';
+  }
+  url += '//';
+  if (aParsedUrl.auth) {
+    url += aParsedUrl.auth + '@';
+  }
+  if (aParsedUrl.host) {
+    url += aParsedUrl.host;
+  }
+  if (aParsedUrl.port) {
+    url += ':' + aParsedUrl.port;
+  }
+  if (aParsedUrl.path) {
+    url += aParsedUrl.path;
+  }
+  return url;
+}
+
+export function isAbsolute(aPath: string) {
+  return aPath.charAt(0) === '/' || urlRegexp.test(aPath);
+}
+
+/**
+ * Normalizes a path, or the path portion of a URL:
+ *
+ * - Replaces consecutive slashes with one slash.
+ * - Removes unnecessary '.' parts.
+ * - Removes unnecessary '<dir>/..' parts.
+ *
+ * Based on code in the Node.js 'path' core module.
+ *
+ * @param aPath The path or url to normalize.
+ */
+export function normalize(aPath: string) {
+  let path = aPath;
+  const url = urlParse(aPath);
+  if (url) {
+    if (!url.path) {
+      return aPath;
+    }
+    path = url.path;
+  }
+  const isAbsoluteBool = isAbsolute(path);
+  // Split the path into parts between `/` characters. This is much faster than
+  // using `.split(/\/+/g)`.
+  const parts = [];
+  let start = 0;
+  let i = 0;
+  while (true) {
+    start = i;
+    i = path.indexOf('/', start);
+    if (i === -1) {
+      parts.push(path.slice(start));
+      break;
+    } else {
+      parts.push(path.slice(start, i));
+      while (i < path.length && path[i] === '/') {
+        i++;
+      }
+    }
+  }
+
+  for (let part, up = 0, i = parts.length - 1; i >= 0; i--) {
+    part = parts[i];
+    if (part === '.') {
+      parts.splice(i, 1);
+    } else if (part === '..') {
+      up++;
+    } else if (up > 0) {
+      if (part === '') {
+        // The first part is blank if the path is absolute. Trying to go
+        // above the root is a no-op. Therefore we can remove all '..' parts
+        // directly after the root.
+        parts.splice(i + 1, up);
+        up = 0;
+      } else {
+        parts.splice(i, 2);
+        up--;
+      }
+    }
+  }
+  path = parts.join('/');
+
+  if (path === '') {
+    path = isAbsoluteBool ? '/' : '.';
+  }
+
+  if (url) {
+    url.path = path;
+    return urlGenerate(url);
+  }
+  return path;
+}
+
+/**
+ * Joins two paths/URLs.
+ *
+ * @param aRoot The root path or URL.
+ * @param aPath The path or URL to be joined with the root.
+ *
+ * - If aPath is a URL or a data URI, aPath is returned, unless aPath is a
+ *   scheme-relative URL: Then the scheme of aRoot, if any, is prepended
+ *   first.
+ * - Otherwise aPath is a path. If aRoot is a URL, then its path portion
+ *   is updated with the result and aRoot is returned. Otherwise the result
+ *   is returned.
+ *   - If aPath is absolute, the result is aPath.
+ *   - Otherwise the two paths are joined with a slash.
+ * - Joining for example 'http://' and 'www.example.com' is also supported.
+ */
+export function join(aRoot: string, aPath: string) {
```

Comment:
This is very similar to `@jridgewell/resolve-uri`'s functionality.

Reply:
I'm trying to integrate it, but see some differences between both libraries.

```
join {
  input: '../../coffee/foo.coffee',
  base: '../temp/temp_maps',
  resolve: '../../coffee/foo.coffee', // @jridgewell/resolve-uri
  join: '../coffee/foo.coffee'// utils.join
}
```

```
sourceFile {
  input: 'temp/bundle.js',
  base: '..',
  resolve: '../temp/bundle.js', // @jridgewell/resolve-uri
  relative: 'temp/bundle.js' // utils.relative
}
```

```
mappedfile {
  input: '../../coffee/foo.coffee',
  base: '..',
  resolve: '../../../coffee/foo.coffee', // @jridgewell/resolve-uri
  relative: '../coffee/foo.coffee' // utils.relative
}
```

I have to admit that I'm not really sure which result is the correct one in those cases.

Reply:
The first is because we need to treat the base as always being a directory and not a file as explained in https://github.com/jridgewell/trace-mapping/blob/main/src/resolve.ts.

The second and third seem to be `relative`, which `resolve-uri` can't currently perform. I'll try adding that functionality this weekend.

Reply:
I've got a working [`relative`](https://github.com/jridgewell/resolve-uri/blob/relative/src/resolve-uri.ts#L244-L288), but still need to add more tests. While going though this, I realized the one provided by `source-map` is really bad. Eg, `relative('a/b/c', 'c/b/a') === 'c/b/a'`, which is just plain wrong.

Reply:
Hi @jridgewell

I'm back from vacation and ready to continue on this PR :D . Is there anything I can help on the `resolve-uri` package ?

Reply:
I'll get the `relativeUri` done this today or tomorrow for more testing here.

Reply:
merged your PR, resolving as  it's working well :)